tests/checkpoints_test.py

```diff
@@ -148,6 +205,69 @@ def test_save_restore_checkpoints_target_none(self):
     expected_new_object = {str(k): v for k, v in enumerate(test_object1)}
     jtu.check_eq(new_object, expected_new_object)
 
+  def test_convert_pre_linen(self):
+    params = checkpoints.convert_pre_linen({
+        'mod_0': {
+            'submod1_0': {},
+            'submod2_1': {},
+            'submod1_2': {},
+        },
+        'mod2_2': {
+            'submod2_2_0': {}
+        },
+        'mod2_11': {
+            'submod2_11_0': {}
+        },
+        'mod2_1': {
+            'submod2_1_0': {}
+        },
+    })
+    self.assertDictEqual(
+        core.unfreeze(params), {
+            'mod_0': {
+                'submod1_0': {},
+                'submod1_1': {},
+                'submod2_0': {},
+            },
+            'mod2_0': {
+                'submod2_1_0': {}
+            },
+            'mod2_1': {
+                'submod2_2_0': {}
+            },
+            'mod2_2': {
+                'submod2_11_0': {}
+            },
+        })
+
+  def test_convert_checkpoint(self):
+    inputs = jnp.ones([2, 5, 5, 1])
+    rng = jax.random.PRNGKey(0)
+    # pre-Linen.
+    with flax.nn.stateful() as model_state:
+      y, params = ModelPreLinen.init(rng, inputs)
+    pre_linen_optimizer = flax.optim.GradientDescent(0.1).create(params)
+    train_state = TrainState(
+        optimizer=pre_linen_optimizer, model_state=model_state)
+    state_dict = flax.serialization.to_state_dict(train_state)
+    # Linen.
+    model = Model()
+    variables = model.init(rng, inputs)
+    optimizer = flax.optim.GradientDescent(0.1).create(variables['params'])
+    optimizer = optimizer.restore_state(
+        flax.core.unfreeze(
+            checkpoints.convert_pre_linen(state_dict['optimizer'])))
+    optimizer = optimizer.apply_gradient(variables['params'])
+    batch_stats= flax.core.freeze(flax.traverse_util.unflatten_dict({
```

Comment:
This `Collection` unflatten trick should probably be in the docstring of convert_pre_linnen or be natively supported by convert_pre_linnen.

Reply:
Added comment to the pydoc.

I prefer to keep the function simple and do a single thing. Copying the unflattening from the pydoc seems easier and safer to use than to make the function contract more complicated.